### ISSUE TITLE
Lot/Serial Number Traceability - memory error when having a very large number of moves

### ISSUE THREAD

**Issue by dreispt** [odoo/odoo#30977]
### Impacted versions:
Odoo 11
See also Enterprise contract Ticket #1937027

### Steps to reproduce:

Using a database where Products are moved with Serial numbers, and there are millions of stock moves.
Menu Inventory -> Master Data -> Lots/Serial Numbers
Choose a Serial Number, and on the from click on the "Up/Down Traceability"

### Current behavior:

Out of memory error.
```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo11/odoo/addons/web/controllers/main.py", line 934, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo11/odoo/addons/web/controllers/main.py", line 926, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/api.py", line 687, in call_kw
    return call_kw_model(method, model, args, kwargs)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/api.py", line 672, in call_kw_model
    result = method(recs, *args, **kwargs)
  File "/opt/odoo11/odoo/addons/stock/models/stock_traceability.py", line 374, in get_html
    return self.create({}).with_context(given_context)._get_html()
  File "/opt/odoo11/odoo/addons/stock/models/stock_traceability.py", line 366, in _get_html
    rcontext['lines'] = self.with_context(context).get_lines()
  File "/opt/odoo11/odoo/addons/stock/models/stock_traceability.py", line 111, in get_lines
    ('move_id.returned_move_ids', '=', False),
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/models.py", line 1480, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/models.py", line 3791, in _search
    self._cr.execute(query_str, where_clause_params)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/sql_db.py", line 155, in wrapper
    return f(self, *args, **kwargs)
  File "/opt/odoo11/env/lib/python3.6/site-packages/odoo-11.0-py3.6.egg/odoo/sql_db.py", line 232, in execute
    res = self._obj.execute(query, params)
psycopg2.InternalError: invalid memory alloc request size 1073741824
```

### Expected behavior:

No error, and serial number move history shown.

### Video/Screenshot link (optional):

Initial investigation reveals that an inefficient search operation is being done, and on databases with a large number of stock moves it returns millions of records, and runs out od memory while trying to process them.

The offending code is probably this: https://github.com/odoo/odoo/blob/11.0/addons/stock/models/stock_traceability.py#L107

**Comment by dreispt** [comment]
It seems that the culprit in in [line 111](https://github.com/odoo/odoo/blob/11.0/addons/stock/models/stock_traceability.py#L111).

```
('move_id.returned_move_ids', '=', False)
```

This is returning a huge number of ids, and th2 "IN" clause in the SQL "WHERE" condition can't be built within reasonable memory  limits.

**Comment by pedrobaeza** [comment]
@dreispt one quick solution is to change in PG `work_mem` parameter to something like 32MB:

https://www.postgresql.org/docs/9.4/runtime-config-resource.html

**Comment by dreispt** [comment]
cc @max3903

**Comment by max3903** [comment]
Thanks @pedrobaeza and @dreispt. I will try that and get back to you here.

**Comment by dreispt** [comment]
I think this is a good case where SQL will do better than the ORM and should be used.
I was curious enough to spend some time on this, and came up with the SQL that should be equivalent to the Domain condition being used:

```
select sml.*
from stock_move_line sml
left join stock_location sld on sld.id=sml.location_dest_id
left join stock_move sm on sm.id=sml.move_id
  left join stock_move smr on smr.id=sm.origin_returned_move_id
where sml.lot_id=%d
and sld.usage='internal'
and sm.state='done'
and smr.id IS NULL
```

To have an actual patch, a little more work is needed, but I think this would the the way to go.

**Comment by pedrobaeza** [comment]
An SQL expression should be avoided always, as it bypasses all the ORM extra features, like for example record rules. Your SQL is not complying with that. Optimization should be done at ORM level or tuning the DB with indexes or auto-joins IMO. Check #30350 for an example of analysis for optimizing ORM.

**Comment by max3903** [comment]
@pedrobaeza We are using PostgreSQL 11 behind pgpool. I changed work_mem to 32MB in PostgreSQL nodes and I still get the same error.

**Comment by gdgellatly** [comment]
Just do a 2 stage search to avoid the in clause. Or search and filter, the semantics are simple. Most records dont have returned moves, few records have a single lot, so do the lot first, then filter that.

**Comment by moylop260** [comment]
FYI check page 17 of the following document:
 - https://www.odoo.com/es_ES/slides/slide/how-to-optimize-the-performance-229/pdf_content

![screen shot 2019-02-11 at 3 57 29 pm](https://user-images.githubusercontent.com/6644187/52596382-c709a600-2e15-11e9-90da-4b3a05786d52.png)


Running
`self.env['stock.move.line'].search([('move_id.returned_move_ids', '=', False)])`

The following queries will be ran:
```query
-- RES1
SELECT distinct("origin_returned_move_id") FROM "stock_move" where "origin_returned_move_id" is not null

-- RES2
SELECT "stock_move".id FROM "stock_move" LEFT JOIN "stock_picking" as "stock_move__picking_id" ON ("stock_move"."picking_id" = "stock_move__picking_id"."id")
WHERE (("stock_move"."id" not in ({RES1})) OR "stock_move"."id" IS NULL)
ORDER BY "stock_move__picking_id"."priority" DESC, "stock_move__picking_id"."date" ASC,"stock_move__picking_id"."id" DESC,"stock_move"."sequence" ,"stock_move"."id"

-- RES3
SELECT "stock_move_line".id FROM "stock_move_line" LEFT JOIN "stock_quant_package" as "stock_move_line__result_package_id" ON ("stock_move_line"."result_package_id" = "stock_move_line__result_package_id"."id")
WHERE ("stock_move_line"."move_id" in ({RES2}))
ORDER BY "stock_move_line__result_package_id"."name" DESC, "stock_move_line"."id"
```

I saw that `res2` is retrieving `10,530` ids for my case and used all them in one query for res3
I mean, it is not splited from:
 - https://github.com/odoo/odoo/blob/66efbf2/odoo/sql_db.py#L256-L259

it is not using the max 1,000 from:
  - https://github.com/odoo/odoo/blob/66efbf2b5bf352fd913e30b6db95bdb3c646582f/odoo/sql_db.py#L144

Maybe that is the real bug for cases where a auto_join is defined.


Full postgresql log:
 - https://gist.github.com/moylop260/038944c340ca1bf9471aa641b9303988

**Comment by dreispt** [comment]
I have the info that this was also reported in OPW 1937027.

**Comment by xmo-odoo** [comment]
I'll close this since @dreispt fixed this specific issue in #31086 and #26702 already exists as the more general issue.
